File: bench/c/good/binary_search.c
/**************************************************************************/
/*                                                                        */
/*  The Why platform for program certification                            */
/*                                                                        */
/*  Copyright (C) 2002-2011                                               */
/*                                                                        */
/*    Jean-Christophe FILLIATRE, CNRS & Univ. Paris-sud 11                */
/*    Claude MARCHE, INRIA & Univ. Paris-sud 11                           */
/*    Yannick MOY, Univ. Paris-sud 11                                     */
/*    Romain BARDOU, Univ. Paris-sud 11                                   */
/*                                                                        */
/*  Secondary contributors:                                               */
/*                                                                        */
/*    Thierry HUBERT, Univ. Paris-sud 11  (former Caduceus front-end)     */
/*    Nicolas ROUSSET, Univ. Paris-sud 11 (on Jessie & Krakatoa)          */
/*    Ali AYAD, CNRS & CEA Saclay         (floating-point support)        */
/*    Sylvie BOLDO, INRIA                 (floating-point support)        */
/*    Jean-Francois COUCHOT, INRIA        (sort encodings, hyps pruning)  */
/*    Mehdi DOGGUY, Univ. Paris-sud 11    (Why GUI)                       */
/*                                                                        */
/*  This software is free software; you can redistribute it and/or        */
/*  modify it under the terms of the GNU Lesser General Public            */
/*  License version 2.1, with the special exception on linking            */
/*  described in file LICENSE.                                            */
/*                                                                        */
/*  This software is distributed in the hope that it will be useful,      */
/*  but WITHOUT ANY WARRANTY; without even the implied warranty of        */
/*  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.                  */
/*                                                                        */
/**************************************************************************/

/*@ axiom mean_1 : \forall int x, int y; x <= y => x <= (x+y)/2 <= y */

/* binary_search(t,n,v) search for element v in array t 
   between index 0 and n-1
   array t is assumed sorted in increasing order
   returns an index i between 0 and n-1 where t[i] equals v, 
   or -1 if no element of t is equal to v  
 */

/*@ requires 
  @   n >= 0 && \valid_range(t,0,n-1) &&
  @   \forall int k1, int k2; 0 <= k1 <= k2 <= n-1 => t[k1] <= t[k2]
  @ ensures
  @   (\result >= 0 && t[\result] == v) ||
  @   (\result == -1 && \forall int k; 0 <= k < n => t[k] != v)
  @*/
int binary_search(int* t, int n, int v) {
  int l = 0, u = n-1;
  /*@ invariant 
    @   0 <= l && u <= n-1 && 
    @   \forall int k; 0 <= k < n => t[k] == v => l <= k <= u
    @ variant u-l 
    @*/
  while (l <= u) {
    int m = (l + u) / 2;
    if (t[m] < v) l = m + 1;
    else if (t[m] > v) u = m - 1;
    else return m; 
  }
  return -1;
}
    
/*
Local Variables: 
compile-command: "make binary_search.overflows"
End: 
*/

File: bench/java/good/BinarySearch.java

//@+ ArithOverflow = no

/*@ axiom mean_1 : 
      \forall int x, y; x <= y ==> x <= x+(y-x)/2 <= y; */

class BinarySearch {

    /* binary_search(t,n,v) search for element v in array t 
       between index 0 and n-1
       array t is assumed to be sorted in increasing order
       returns an index i between 0 and n-1 where t[i] equals v, 
       or -1 if no element in t is equal to v  
    */
    
    /*@ requires 
      @   0 <= n <= t.length &&
      @   \forall int k1, k2; 0 <= k1 <= k2 <= n-1 ==> t[k1] <= t[k2];
      @ behavior correctness:
      @ ensures
      @   (\result >= 0 && t[\result] == v) ||
      @   (\result == -1 && \forall int k; 0 <= k < n ==> t[k] != v);
      @*/
    int binary_search(int t[], int n, int v) {
	int l = 0, u = n-1;
	/*@ loop_invariant 
	  @   0 <= l && u <= n-1 && 
	  @   \forall int k; 0 <= k < n ==> t[k] == v ==> l <= k <= u;
	  @ decreases 
	  @   u-l ;
	  @*/
	while (l <= u ) {
	    int m = l + (u - l) / 2;
	    if (t[m] < v) l = m + 1;
	    else if (t[m] > v) u = m - 1;
	    else return m; 
	}
	return -1;
    }

    /* idiomatic code, using t.length instead of n as argument */

    /*@ requires 
      @   t != null &&
      @   \forall int k1, k2; 0 <= k1 <= k2 <= t.length-1 ==> t[k1] <= t[k2];
      @ behavior correctness:
      @ ensures
      @   (\result >= 0 && t[\result] == v) ||
      @   (\result == -1 && \forall int k; 0 <= k < t.length ==> t[k] != v);
      @*/
    int binary_search2(int t[], int v) {
	int l = 0, u = t.length - 1;
	/*@ loop_invariant 
	  @   0 <= l && u <= t.length - 1 && 
	  @   \forall int k; 0 <= k < t.length ==> t[k] == v ==> l <= k <= u;
	  @ decreases 
	  @   u-l ;
	  @*/
	while (l <= u ) {
	    int m = l + (u - l) / 2;
	    if (t[m] < v) l = m + 1;
	    else if (t[m] > v) u = m - 1;
	    else return m; 
	}
	return -1;
    }

}
    
/*
Local Variables: 
compile-command: "make BinarySearch.io"
End: 
*/
